lisa/sut_orchestrator/azure/tools.py

```diff
@@ -81,16 +85,50 @@ def deprovision(self) -> None:
         self.run("-deprovision --force", sudo=True, expected_exit_code=0)
 
     def upgrade_from_source(self) -> None:
-        git = self.node.tools[Git]
-        git.clone(self._src_url, cwd=self.node.working_path)
+        # Pin the WALinuxAgent to v2.9.0.4 because the latest v2.9.1.1
+        # fails to assign IP address correctly to ib interface
+        wget = self.node.tools[Wget]
+        tarball_name = "v2.9.0.4.tar.gz"
+        walinuxagent_download_url = (
+            f"https://github.com/Azure/WALinuxAgent/archive/refs/tags"
+            f"/{tarball_name}"
+        )
+
+        try:
+            wget.get(
+                url=walinuxagent_download_url,
+                file_path=str(self.node.working_path),
+                filename=tarball_name,
+                overwrite=False,
+                sudo=True,
+            )
+        except LisaException as identifier:
+            if "404: Not Found." in str(identifier):
+                raise UnsupportedDistroException(
+                    self.node.os, f"{walinuxagent_download_url} doesn't exist."
+                )
+
+        tar = self.node.tools[Tar]
+        tar.extract(
+            file=f"{self.node.working_path}/{tarball_name}",
+            dest_dir=str(self.node.working_path),
+            gzip=True,
+            sudo=True,
+        )
+
         python_cmd, _ = self.get_python_cmd()
         for package in list(["python-setuptools", "python3-setuptools"]):
             if self.node.os.is_package_in_repo(package):  # type: ignore
                 self.node.os.install_packages(package)  # type: ignore
+
+        if not self.node.tools[Ls].path_exists(path="/usr/bin/python"):
```

Comment:
these logic should be put in self.get_python_cmd

Reply:
Moved to this function